### PR TITLE
Bug 2047732: [IBM]Volume is not deleted after destroy cluster

### PR DESCRIPTION
**Pull request by sameshai**
**Description**

The existing GCP destroy disk logic will continue to execute listdisks() until there are zero pending items in the list i.e until all disks are deleted successfully. We found this logic as redundant as there is no point in calling list disks as we are in uninstallation phase. There is no chance of new disk coming into picture. 

Hence we have come-up with combination of both the flows. Note that we will be running infinitely until all the items in the list are deleted successfully. The difference is we are avoiding unnecessary listDisks() calls like GCP and making sure in the same first call to destroy disk we will wait for all disks to go away. If still something is pending it will be retried from the pending list.

1.) We will list all disks using pagination and filter it based on tags. 
2.) The list of disks in #1 will be added to pendingItem list.
3.) Disk deletion will be invoked for all the disks and if any of the call fails it will return error and retried later with pendingItem list logic.
4.) Once all disk deletion is invoked we will wait for the all the disk deletion calls to get completed. Ideally first few disk deletion will take time , rest of others would have been already deleted. 
5.) The pendingItem list will be empty only if all disk deletion would have been deleted successfully. 
6.) If there is any issue in disk deletion for around 20 mins user will have to take action as suggested in the error for respective disks.

`time="2022-06-10T12:20:22Z" level=debug msg="Failed to delete disk name=pvc-0d111d56-cf74-4045-8df0-fc439b1b4cb8, id=r018-ef565808-3491-4262-a6d5-27428e27ec0b.If this error continues to persist for more than 20 minutes then please try to manually cleanup the volume using - ibmcloud is vol r018-ef565808-3491-4262-a6d5-27428e27ec0b: Delete volume failed. Volume can be deleted only when its status is available or failed."`

### CODE REVIEW

File: pkg/destroy/ibmcloud/disk.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	options := o.vpcSvc.NewDeleteVolumeOptions(item.id)
+	details, err := o.vpcSvc.DeleteVolumeWithContext(ctx, options)
+
+	if err != nil {
+		return errors.Wrapf(err, "Failed to delete volume %q", item.name)
+	}
+
+	var skip = false
+
+	if err == nil && details != nil && details.StatusCode == http.StatusNoContent {
```
Comment (GunaKKIBM) [inline pkg/destroy/ibmcloud/disk.go:54]: It is already returning at line 48, if err != nil, I think, it's ok to remove err==nil condition from line 54.

File: pkg/destroy/ibmcloud/ibmcloud.go [hunk 2]
```diff
@@ -403,3 +438,11 @@ func (t pendingItemTracker) deletePendingItems(itemType string, items []cloudRes
 func isErrorStatus(code int64) bool {
 	return code != 0 && (code < 200 || code >= 300)
 }
+
+func (o *ClusterUninstaller) clusterIDFilter() string {
+	return fmt.Sprintf("clusterid:\"%s-*\"", o.InfraID)
+}
+
+func (o *ClusterUninstaller) clusterLabelFilter() string {
+	return fmt.Sprintf("kubernetes-io-cluster-%s:\"owned\"", o.InfraID)
```
Comment (GunaKKIBM) [inline pkg/destroy/ibmcloud/ibmcloud.go:447]: Just want to know, for this label, if double quotes are required ?

File: pkg/destroy/ibmcloud/disk.go (hunk 1 above)
Comment (sameshai) [reply pkg/destroy/ibmcloud/disk.go:54]: ok I am hold on this as we might need to change the logic after todays discussion.

File: pkg/destroy/ibmcloud/disk.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// destroyDisks removes all disk resources that have a name prefixed
+// with the cluster's infra ID.
+func (o *ClusterUninstaller) destroyDisks() error {
+	found, err := o.listDisks()
+	if err != nil {
+		return err
+	}
+	items := o.insertPendingItems("disk", found)
```
[the remark this answers is not in the document]
Comment (sameshai) [reply pkg/destroy/ibmcloud/disk.go:90]: It will do nothing it will just return empty list of items for disk. Please check the logic of insertPendingItems.

File: go.mod [hunk 4]
```diff
@@ -12,10 +12,10 @@ require (
 	github.com/IBM-Cloud/bluemix-go v0.0.0-20211102075456-ffc4e11dfb16
 	github.com/IBM-Cloud/power-go-client v1.1.5
 	github.com/IBM/go-sdk-core/v4 v4.9.0
-	github.com/IBM/go-sdk-core/v5 v5.9.1
+	github.com/IBM/go-sdk-core/v5 v5.9.5
 	github.com/IBM/networking-go-sdk v0.14.0
 	github.com/IBM/platform-services-go-sdk v0.18.16
-	github.com/IBM/vpc-go-sdk v1.0.1
+	github.com/IBM/vpc-go-sdk v0.20.0
```
Comment (r4f4) [inline go.mod:18]: Is the downgrade needed? Is quite a version jump.
Also vendor changes should be done in their own commit.

Comment (sameshai) [reply go.mod:18]: @r4f4 this is not downgrade they have specified this as new version

Comment (r4f4) [reply go.mod:18]: Indeed [1], that's confusing.

[1] https://github.com/IBM/vpc-go-sdk#ibm-cloud-vpc-go-sdk-version-0200

File: pkg/destroy/ibmcloud/disk.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	o.Logger.Infof("Deleting disk %s", item.name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	options := o.vpcSvc.NewDeleteVolumeOptions(item.id)
+	details, err := o.vpcSvc.DeleteVolumeWithContext(ctx, options)
+
+	if err != nil && details.StatusCode == http.StatusNotFound {
+		// The resource is gone
+		o.deletePendingItems(item.typeName, []cloudResource{item})
+		o.Logger.Infof("Deleted disk %s", item.name)
```
Comment (GunaKKIBM) [inline pkg/destroy/ibmcloud/disk.go:51]: Considering the disk is deleted here, is it ok to return from here ?

Comment (sameshai) [reply pkg/destroy/ibmcloud/disk.go:55]: We can but we are doing return nil once at the end for success case.

File: pkg/destroy/ibmcloud/disk.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+	"github.com/pkg/errors"
+)
+
+func (o *ClusterUninstaller) listDisks() ([]cloudResource, error) {
+	o.Logger.Infof("Listing disks")
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	options := o.vpcSvc.NewListVolumesOptions()
```
Comment (sameshai) [inline pkg/destroy/ibmcloud/disk.go:15]: 1.) Should we consider Pagination here ?

**Review by arahamad — COMMENTED** [review]
I see some gap and expectation

1- pagination, lets say we have 1000+ volume which needs to be deleted but list API may not return all so there will be partial volume deletion
2- why we are calling list api again and again as this volume list will not going to change for cluster which is going to be deleted

**Comment by sameshai** [comment]
> I see some gap and expectation
> 
> 1- pagination, lets say we have 1000+ volume which needs to be deleted but list API may not return all so there will be partial volume deletion 2- why we are calling list api again and again as this volume list will not going to change for cluster which is going to be deleted

Pagination is implemented on list volume again and again we need to discuss with @cjschaef as this how all the deletions are implemented. We can conclude on the call.

**Comment by sameshai** [comment]
> I see some gap and expectation
> 
> 1- pagination, lets say we have 1000+ volume which needs to be deleted but list API may not return all so there will be partial volume deletion 2- why we are calling list api again and again as this volume list will not going to change for cluster which is going to be deleted

@arahamad All tested and verified please review
```

time="2022-06-14T10:36:16Z" level=debug msg="Listing virtual service instances"
time="2022-06-14T10:36:16Z" level=info msg="Listing disks"
time="2022-06-14T10:36:17Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-master-2\""
time="2022-06-14T10:36:18Z" level=debug msg="Found disk: r018-e212a649-3a98-43a5-9246-b7fc631d8e96"
time="2022-06-14T10:36:18Z" level=debug msg="Found disk: r018-e091ca35-8cd5-44aa-b044-c60cece06815"
time="2022-06-14T10:36:18Z" level=debug msg="Listing next page r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82"
time="2022-06-14T10:36:18Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-master-1\""
time="2022-06-14T10:36:19Z" level=debug msg="Found disk: r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82"
time="2022-06-14T10:36:19Z" level=debug msg="Found disk: r018-7dbc38f8-0b94-45ba-b81c-9b7f5a336860"
time="2022-06-14T10:36:19Z" level=debug msg="Listing next page r018-44f19222-3b21-4536-be37-a018192c96a1"
time="2022-06-14T10:36:20Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-master-0\""
time="2022-06-14T10:36:20Z" level=debug msg="Found disk: r018-44f19222-3b21-4536-be37-a018192c96a1"
time="2022-06-14T10:36:20Z" level=debug msg="Found disk: r018-1bc4ff07-7da3-4efa-b7fb-410ebea2b67c"
time="2022-06-14T10:36:20Z" level=debug msg="Listing next page r018-4f7afc43-cb33-4433-9c03-67d16ff9aae5"
time="2022-06-14T10:36:20Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-worker-3-72glp\""
time="2022-06-14T10:36:20Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-worker-2-xbd6m\""
time="2022-06-14T10:36:20Z" level=debug msg="Deleting instance \"sam-ipi-hope-kmgjz-worker-1-s64lm\""
time="2022-06-14T10:36:21Z" level=debug msg="Instances: 6 items pending"
time="2022-06-14T10:36:21Z" level=debug msg="Listing next page r018-360b0847-2d84-4510-874b-d8644aa26e0f"
time="2022-06-14T10:36:22Z" level=debug msg="Listing next page r018-ee7e7ba6-5714-4f8b-8130-fcc827c18906"
time="2022-06-14T10:36:23Z" level=debug msg="Listing next page r018-8cc35b19-eea8-408e-bea4-0035383fecd5"
time="2022-06-14T10:36:24Z" level=debug msg="Listing next page r018-d7e1997e-f694-45ce-a672-351c5537cf18"
time="2022-06-14T10:36:25Z" level=debug msg="Listing next page r018-b0d9dc3c-8458-4db9-b94e-37ea289b539b"
time="2022-06-14T10:36:26Z" level=debug msg="Listing next page r018-080f9c2d-7468-4a0f-9bf2-60d4e40df51f"
time="2022-06-14T10:36:27Z" level=debug msg="All disks fetched"
time="2022-06-14T10:36:27Z" level=info msg="Deleting disk r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82"
time="2022-06-14T10:36:28Z" level=info msg="Disk deletion was initiated, disk r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82"
time="2022-06-14T10:36:28Z" level=info msg="Deleting disk r018-7dbc38f8-0b94-45ba-b81c-9b7f5a336860"
time="2022-06-14T10:36:29Z" level=info msg="Disk deletion was initiated, disk r018-7dbc38f8-0b94-45ba-b81c-9b7f5a336860"
time="2022-06-14T10:36:29Z" level=info msg="Deleting disk r018-44f19222-3b21-4536-be37-a018192c96a1"
time="2022-06-14T10:36:30Z" level=info msg="Disk deletion was initiated, disk r018-44f19222-3b21-4536-be37-a018192c96a1"
time="2022-06-14T10:36:30Z" level=info msg="Deleting disk r018-1bc4ff07-7da3-4efa-b7fb-410ebea2b67c"
time="2022-06-14T10:36:30Z" level=info msg="Disk deletion was initiated, disk r018-1bc4ff07-7da3-4efa-b7fb-410ebea2b67c"
time="2022-06-14T10:36:30Z" level=info msg="Deleting disk r018-e212a649-3a98-43a5-9246-b7fc631d8e96"
time="2022-06-14T10:36:31Z" level=debug msg="Listing virtual service instances"
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-worker-1-s64lm\""
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-master-2\""
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-master-1\""
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-master-0\""
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-worker-3-72glp\""
time="2022-06-14T10:36:31Z" level=info msg="Deleted instance \"sam-ipi-hope-kmgjz-worker-2-xbd6m\""
time="2022-06-14T10:36:31Z" level=info msg="Disk deletion was initiated, disk r018-e212a649-3a98-43a5-9246-b7fc631d8e96"
time="2022-06-14T10:36:31Z" level=info msg="Deleting disk r018-e091ca35-8cd5-44aa-b044-c60cece06815"
time="2022-06-14T10:36:32Z" level=info msg="Disk deletion was initiated, disk r018-e091ca35-8cd5-44aa-b044-c60cece06815"
time="2022-06-14T10:36:32Z" level=info msg="Waiting for disk r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82 to be deleted"
time="2022-06-14T10:36:33Z" level=info msg="UNEXPECTED RESULT, Re-attempting execution .., attempt=1, retry-gap=10, max-retry-Attempts=10, stopRetry=false, error=<nil>"
time="2022-06-14T10:36:45Z" level=info msg="UNEXPECTED RESULT, Re-attempting execution .., attempt=2, retry-gap=20, max-retry-Attempts=10, stopRetry=false, error=<nil>"
time="2022-06-14T10:37:05Z" level=info msg="UNEXPECTED RESULT, Re-attempting execution .., attempt=3, retry-gap=40, max-retry-Attempts=10, stopRetry=false, error=<nil>"
time="2022-06-14T10:37:45Z" level=info msg="Deleted disk r018-d14f5b9a-22b3-4efd-9750-a6cf13a1ee82"
time="2022-06-14T10:37:45Z" level=info msg="Waiting for disk r018-7dbc38f8-0b94-45ba-b81c-9b7f5a336860 to be deleted"
time="2022-06-14T10:37:46Z" level=info msg="Deleted disk r018-7dbc38f8-0b94-45ba-b81c-9b7f5a336860"
time="2022-06-14T10:37:46Z" level=info msg="Waiting for disk r018-44f19222-3b21-4536-be37-a018192c96a1 to be deleted"
time="2022-06-14T10:37:46Z" level=info msg="Deleted disk r018-44f19222-3b21-4536-be37-a018192c96a1"
time="2022-06-14T10:37:46Z" level=info msg="Waiting for disk r018-1bc4ff07-7da3-4efa-b7fb-410ebea2b67c to be deleted"
time="2022-06-14T10:37:46Z" level=info msg="Deleted disk r018-1bc4ff07-7da3-4efa-b7fb-410ebea2b67c"
time="2022-06-14T10:37:46Z" level=info msg="Waiting for disk r018-e212a649-3a98-43a5-9246-b7fc631d8e96 to be deleted"
time="2022-06-14T10:37:46Z" level=info msg="Deleted disk r018-e212a649-3a98-43a5-9246-b7fc631d8e96"
time="2022-06-14T10:37:46Z" level=info msg="Waiting for disk r018-e091ca35-8cd5-44aa-b044-c60cece06815 to be deleted"
time="2022-06-14T10:37:46Z" level=info msg="Deleted disk r018-e091ca35-8cd5-44aa-b044-c60cece06815"
time="2022-06-14T10:37:46Z" level=debug msg="Listing load balancers"
time="2022-06-14T10:37:49Z" level=debug msg="Deleting load balancer \"kube-sam-ipi-hope-kmgjz-69f02044e8804dca81b0268e02f8df87\""
time="2022-06-14T10:38:00Z" level=debug msg="Deleting load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-private\""
time="2022-06-14T10:38:12Z" level=debug msg="Deleting load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-public\""
time="2022-06-14T10:38:23Z" level=debug msg="Load Balancers: 3 items pending"
time="2022-06-14T10:38:33Z" level=debug msg="Listing load balancers"
time="2022-06-14T10:38:36Z" level=debug msg="Waiting for load balancer \"kube-sam-ipi-hope-kmgjz-69f02044e8804dca81b0268e02f8df87\" to delete"
time="2022-06-14T10:38:36Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-private\" to delete"
time="2022-06-14T10:38:36Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-public\" to delete"
time="2022-06-14T10:38:36Z" level=debug msg="Load Balancers: 3 items pending"
time="2022-06-14T10:38:43Z" level=debug msg="Listing load balancers"
time="2022-06-14T10:38:49Z" level=debug msg="Waiting for load balancer \"kube-sam-ipi-hope-kmgjz-69f02044e8804dca81b0268e02f8df87\" to delete"
time="2022-06-14T10:38:49Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-private\" to delete"
time="2022-06-14T10:38:49Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-public\" to delete"
time="2022-06-14T10:38:49Z" level=debug msg="Load Balancers: 3 items pending"
time="2022-06-14T10:38:53Z" level=debug msg="Listing load balancers"
time="2022-06-14T10:38:56Z" level=debug msg="Waiting for load balancer \"kube-sam-ipi-hope-kmgjz-69f02044e8804dca81b0268e02f8df87\" to delete"
time="2022-06-14T10:38:56Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-private\" to delete"
time="2022-06-14T10:38:56Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-public\" to delete"
time="2022-06-14T10:38:56Z" level=debug msg="Load Balancers: 3 items pending"
time="2022-06-14T10:39:03Z" level=debug msg="Listing load balancers"
time="2022-06-14T10:39:06Z" level=debug msg="Waiting for load balancer \"kube-sam-ipi-hope-kmgjz-69f02044e8804dca81b0268e02f8df87\" to delete"
time="2022-06-14T10:39:06Z" level=info msg="Deleted load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-private\""
time="2022-06-14T10:39:06Z" level=debug msg="Waiting for load balancer \"sam-ipi-hope-kmgjz-kubernetes-api-public\" to delete"
time="2022-06-14T10:39:06Z" level=debug msg="Load Balancers: 2 items pending"
time="2022-06-14T10:39:13Z" level=debug msg="Listing load balancers"
```

**Review by GunaKKIBM — APPROVED** [review]

**Review by GunaKKIBM — APPROVED** [review]

**Review by jstuever — COMMENTED** [review]
The PR look functionally ok. However, I suspect there is some unnecessary code here. I see this borrows heavily from the GCP destroy. One significant difference is that GCP destroy runs in an infinite loop until everything is deleted whereas this code runs only one time with loops on individual items. I see it is using the pendingItems() functionality, which I'm not sure makes (as much) sense without that primary loop. However, this PR didn't add this pattern, so I don't feel we should block on this.

Please add a description to the PR.

File: pkg/destroy/ibmcloud/disk.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	result := []cloudResource{}
+	clusterOwnedTag := o.clusterLabelFilter()
+	options := o.vpcSvc.NewListVolumesOptions()
+
+	for {
+		ctx, cancel := o.contextWithTimeout()
+		defer cancel()
+		options.SetLimit(100)
+		resources, _, err := o.vpcSvc.ListVolumesWithContext(ctx, options)
+		if err != nil {
```
Comment (arahamad) [inline pkg/destroy/ibmcloud/disk.go:22]: in case of error for any page we are returning nil, is this expected i.e no volume deletion in case of error or do we need to return the partial list or need to to re-try

**Review by arahamad — COMMENTED** [review]
please check these comments and take action accordingly

File: pkg/destroy/ibmcloud/disk.go (hunk 7 above)
Comment (sameshai) [reply pkg/destroy/ibmcloud/disk.go:22]: Lets discuss with IPI team on this

File: pkg/destroy/ibmcloud/disk.go (hunk 1 above)
Comment (sameshai) [reply pkg/destroy/ibmcloud/disk.go:71]: Done

File: pkg/destroy/ibmcloud/ibmcloud.go [hunk 8]
```diff
@@ -74,9 +76,41 @@ func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.
...
+		if i > 0 {
+			time.Sleep(time.Duration(retryGap) * time.Second)
+		}
+		// Call function which required retry, retry is decided by function itself
+		err, stopRetry = funcToRetry()
+		if stopRetry {
+			break
+		}
+
+		// Update retry gap as per exponentioal
```
[the remark this answers is not in the document]
Comment (sameshai) [reply pkg/destroy/ibmcloud/ibmcloud.go:99]: I did follow the library flow we do it exponential for delete volume wait but I have no issue in doing it with constant retry , I am removing that and put max attempts as 30 so total 5 mins we will wait for volume to delete else it will be retried by the existing installer framework using the pendingdiskitems logic.

**Comment by sameshai** [comment]
> The PR look functionally ok. However, I suspect there is some unnecessary code here. I see this borrows heavily from the GCP destroy. One significant difference is that GCP destroy runs in an infinite loop until everything is deleted whereas this code runs only one time with loops on individual items. I see it is using the pendingItems() functionality, which I'm not sure makes (as much) sense without that primary loop. However, this PR didn't add this pattern, so I don't feel we should block on this.
> 
> Please add a description to the PR.

@jstuever thanks for the quick review. Apologies missed the PR description

**Comment by MayXuQQ** [comment]
@chao007 pre-merge [build](https://storage.googleapis.com/origin-ci-test/logs/release-openshift-origin-installer-launch-gcp-modern/1539060222482452480/build-log.txt)

**Comment by MayXuQQ** [comment]
pre-merge test OK,
after add label such as 'kubernetes-io-cluster-zhsunibm-j68d5:owned' on volume, destroy the cluster, just remains the subnet resource. 
/label qe-approved
/bugzilla cc-qa

**Comment by sameshai** [comment]
@r4f4 can we go ahead with merge

**Review by arahamad — APPROVED** [review]
lgtm, only I see issue at https://github.com/openshift/installer/pull/5962/files#diff-c98fece70f92d74f8db64396e4256d51b5ff7c1861479479360410ec388556e3R23 line where we are returning which will cause issue as any page in that re-try fail then we are marking entire list failed and this loop will keep going on.

**Comment by sameshai** [comment]
> lgtm, only I see issue at https://github.com/openshift/installer/pull/5962/files#diff-c98fece70f92d74f8db64396e4256d51b5ff7c1861479479360410ec388556e3R23 line where we are returning which will cause issue as any page in that re-try fail then we are marking entire list failed and this loop will keep going on.

@arahamad I have already mentioned in the slack channel discussion.So are following GCP flow as of now. 

https://coreos.slack.com/archives/C01U40AM37F/p1655735741598449?thread_ts=1655372841.635669&cid=C01U40AM37F
